### PR TITLE
Fix radio button toggle behavior for keyboard users

### LINKED ISSUE #16223: tab key select radio button in buttons group in form submits no data
reduced test case: 
http://jsfiddle.net/L9fep626/3/

This is a sample form I created using radio buttons following the example in http://getbootstrap.com/javascript/#buttons-checkbox-radio

What I found weird is that, if I 
1. type in recipient text input and message textarea, and then 
2. use [tab] key to jump to the button-group, 
press left/right button and space as one would expect to select a radio button, then 
3. click submit, 
4. result: on the receiving form, the $_POST shows no submission data about the radio button.

This is in contrast to using mouse to select radio button, the selected radio button will be captured and shown in $_POST data.

I would assume, the latter one is the expected behaviour of selecting a button even using keyboard. I have put in a console.log() at onchange event of each radio input, and both keyboard and mouse show that button checked is true.

This is tested on current version of Chrome and Firefox on Ubuntu.

Is this a misbehaviour due to Bootstrap code?
(http://jsfiddle.net/3j2nt0hz/1/ exactly the same html but disabling bootstrap, it works both using keyboard tab-arrowing and mouse clicking).

I would like to investigate what's going on when I [tab] and press arrow key, as contrast to clicking the mouse. How do I do that?

### PR DESCRIPTION
**Pull request by patrickhlauke**
as this breaks keyboard navigation for radio button toggles - fixes https://github.com/twbs/bootstrap/issues/16223

### CODE REVIEW

**Comment by twbs-savage** [comment]
**Tests passed.** Automated cross-browser testing via Sauce Labs and Travis CI shows that the JavaScript changes in this pull request are: **CONFIRMED**

Commit: f6531aa68d51085eb5220ee0e18b9f6b7b81fafd
Build details: https://travis-ci.org/twbs-savage/bootstrap/builds/57098788

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

**Comment by cvrebert** [comment]
This needs a unit test.

**Comment by cvrebert** [comment]
Also, is there a reason the condition isn't `.is('input[type="radio"]')` ?

**Comment by patrickhlauke** [comment]
Not sure how to unit test "if a radio button has been chosen via the keyboard", and not quite sure what to actually test for...thoughts?

As for ``.is('input[type="radio"]')`` ... well yes, that's a bit more terse (and it shows that I don't usually jQuery but do vanilla JS)

**Comment by twbs-savage** [comment]
**Tests passed.** Automated cross-browser testing via Sauce Labs and Travis CI shows that the JavaScript changes in this pull request are: **CONFIRMED**

Commit: 3c8977f0560bd620143ca032198e3b3d44d4b3ce
Build details: https://travis-ci.org/twbs-savage/bootstrap/builds/57108389

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

**Comment by patrickhlauke** [comment]
@cvrebert any idea/pointer for how this could be unit-tested? simulate actual keyboard interactions somehow?

**Comment by cvrebert** [comment]
This doesn't actually seem to fix anything: Demo: http://jsfiddle.net/oxnksxaj/

**Comment by twbs-lmvtfy** [comment]
Hi @cvrebert!

You appear to have posted a live example (http://fiddle.jshell.net/oxnksxaj/show/light/), which is always a good first step. However, according to [the HTML5 validator](http://html5.validator.nu), **your example has some validation errors**, which might potentially be causing your issue:
* line 156, column 9: Attribute ```` autocomplete ```` not allowed on element ```` input ```` at this point.
* line 160, column 29: Attribute ```` autocomplete ```` not allowed on element ```` input ```` at this point.
* line 165, column 29: Attribute ```` autocomplete ```` not allowed on element ```` input ```` at this point.

You'll need to **fix these errors** and post a revised example before we can proceed further.
Thanks!

(*Please note that this is a [fully automated](https://github.com/cvrebert/lmvtfy) comment.*)

File: js/button.js [hunk 1]
```diff
@@ -107,7 +107,7 @@
       var $btn = $(e.target)
       if (!$btn.hasClass('btn')) $btn = $btn.closest('.btn')
```
Comment (cvrebert) [inline js/button.js]: Pretty sure your check needs to happen before this line; otherwise, `$btn` will *always* be the `label.btn`.

**Comment by twbs-savage** [comment]
**Tests passed.** Automated cross-browser testing via Sauce Labs and Travis CI shows that the JavaScript changes in this pull request are: **CONFIRMED**

Commit: 613eedd0db8cacd387e3747a13a3d2acff8e69fe
Build details: https://travis-ci.org/twbs-savage/bootstrap/builds/57940169

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

**Comment by patrickhlauke** [comment]
@cvrebert not sure what the heck i did there...i could have sworn it worked (but then i did mess around with the JS/test a bit too much and probably lost track at some point of what i was doing).

Fixed, i think...see http://jsfiddle.net/vjjmj4da/1/

**Comment by twbs-lmvtfy** [comment]
Hi @patrickhlauke!

You appear to have posted a live example (http://fiddle.jshell.net/vjjmj4da/1/show/light/), which is always a good first step. However, according to [the HTML5 validator](http://html5.validator.nu), **your example has some validation errors**, which might potentially be causing your issue:
* line 155, column 9: Attribute ```` autocomplete ```` not allowed on element ```` input ```` at this point.
* line 159, column 29: Attribute ```` autocomplete ```` not allowed on element ```` input ```` at this point.
* line 164, column 29: Attribute ```` autocomplete ```` not allowed on element ```` input ```` at this point.

You'll need to **fix these errors** and post a revised example before we can proceed further.
Thanks!

(*Please note that this is a [fully automated](https://github.com/cvrebert/lmvtfy) comment.*)

File: js/button.js (hunk 1 above)
Comment (patrickhlauke) [inline js/button.js]: yup, absolutely right. amended below by referencing `e.target` instead

**Comment by cvrebert** [comment]
@patrickhlauke Okay, but now it suffers from the "double `change` events" problem which is the reason that the `preventDefault` was added in the first place.
Open http://jsfiddle.net/vjjmj4da/2/ , change the selected option, and watch the console output.

**Comment by twbs-lmvtfy** [comment]
Hi @cvrebert!

You appear to have posted a live example (http://fiddle.jshell.net/vjjmj4da/2/show/light/), which is always a good first step. However, according to [the HTML5 validator](http://html5.validator.nu), **your example has some validation errors**, which might potentially be causing your issue:
* line 158, column 9: Attribute ```` autocomplete ```` not allowed on element ```` input ```` at this point.
* line 162, column 29: Attribute ```` autocomplete ```` not allowed on element ```` input ```` at this point.
* line 167, column 29: Attribute ```` autocomplete ```` not allowed on element ```` input ```` at this point.

You'll need to **fix these errors** and post a revised example before we can proceed further.
Thanks!

(*Please note that this is a [fully automated](https://github.com/cvrebert/lmvtfy) comment.*)

**Comment by cvrebert** [comment]
Regarding testing this, FWICT, the relevant difference between mouse and keyboard is that keyboard usage results in click events fired at the invisible `<input>`s whereas mouse usage results in click events fired at the `<label>`s.

**Comment by patrickhlauke** [comment]
for the double `change` event, i *may* have found a way around it. just posting here as a fiddle for now http://jsfiddle.net/203pxaej/ - basically, had to mess around with the actual `toggle` function as well

```
if ($input.prop('type') == 'radio') {
  if ($input.prop('checked')) changed = false
  if (!$input.prop('checked') || !this.$element.hasClass('active')) $parent.find('.active').removeClass('active')
}
```

from cursory testing, this now seems to correctly address mouse and keyboard usage, unless i've missed something.

thanks for the info on testing. will have a go at making some tests later too.

**Comment by twbs-lmvtfy** [comment]
Hi @patrickhlauke!

You appear to have posted a live example (http://fiddle.jshell.net/203pxaej/show/light/), which is always a good first step. However, according to [the HTML5 validator](http://html5.validator.nu), **your example has some validation errors**, which might potentially be causing your issue:
* line 158, column 9: Attribute ```` autocomplete ```` not allowed on element ```` input ```` at this point.
* line 162, column 29: Attribute ```` autocomplete ```` not allowed on element ```` input ```` at this point.
* line 167, column 29: Attribute ```` autocomplete ```` not allowed on element ```` input ```` at this point.

You'll need to **fix these errors** and post a revised example before we can proceed further.
Thanks!

(*Please note that this is a [fully automated](https://github.com/cvrebert/lmvtfy) comment.*)

**Comment by twbs-savage** [comment]
**Tests failed.** Automated cross-browser testing via Sauce Labs and Travis CI shows that the JavaScript changes in this pull request are: **BUSTED**

Commit: 8b41c977a38ac191a595647a6f1814e86e90f31c
Build details: https://travis-ci.org/twbs-savage/bootstrap/builds/58083444

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

**Comment by cvrebert** [comment]
@twbs-savage Retry

**Comment by twbs-savage** [comment]
**Tests passed.** Automated cross-browser testing via Sauce Labs and Travis CI shows that the JavaScript changes in this pull request are: **CONFIRMED**

Commit: 8b41c977a38ac191a595647a6f1814e86e90f31c
Build details: https://travis-ci.org/twbs-savage/bootstrap/builds/58084537

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

File: js/button.js [hunk 2]
```diff
@@ -56,8 +56,8 @@
     if ($parent.length) {
       var $input = this.$element.find('input')
       if ($input.prop('type') == 'radio') {
-        if ($input.prop('checked') && this.$element.hasClass('active')) changed = false
-        else $parent.find('.active').removeClass('active')
+          if ($input.prop('checked')) changed = false
```
Comment (cvrebert) [inline js/button.js]: This is indented one level too much

**Comment by twbs-savage** [comment]
**Tests passed.** Automated cross-browser testing via Sauce Labs and Travis CI shows that the JavaScript changes in this pull request are: **CONFIRMED**

Commit: 1b9baa0d98019eba187e65102959fff60d06538a
Build details: https://travis-ci.org/twbs-savage/bootstrap/builds/58088920

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

**Comment by twbs-savage** [comment]
**Tests passed.** Automated cross-browser testing via Sauce Labs and Travis CI shows that the JavaScript changes in this pull request are: **CONFIRMED**

Commit: 66d71136e668cf7bc8a311299fad9b773b76285a
Build details: https://travis-ci.org/twbs-savage/bootstrap/builds/58089184

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

**Comment by patrickhlauke** [comment]
> Regarding testing this, FWICT, the relevant difference between mouse and keyboard is that keyboard usage results in `click` events fired at the invisible `<input>`s whereas mouse usage results in click events fired at the `<label>`s.

Unfortunately, this doesn't seem to be the case - even testing the case that is currently broken for keyboard users http://jsfiddle.net/oxnksxaj/ and explictly doing something like

```
$("#discover").trigger('click')
```

in the console does check the radio button correctly, which is not the case for actual keyboard use. I might be wrong, but I can't see any immediate way to test/simulate keyboard interaction in this case?

**Comment by twbs-lmvtfy** [comment]
Hi @patrickhlauke!

You appear to have posted a live example (http://fiddle.jshell.net/oxnksxaj/show/light/), which is always a good first step. However, according to [the HTML5 validator](http://html5.validator.nu), **your example has some validation errors**, which might potentially be causing your issue:
* line 156, column 9: Attribute ```` autocomplete ```` not allowed on element ```` input ```` at this point.
* line 160, column 29: Attribute ```` autocomplete ```` not allowed on element ```` input ```` at this point.
* line 165, column 29: Attribute ```` autocomplete ```` not allowed on element ```` input ```` at this point.

You'll need to **fix these errors** and post a revised example before we can proceed further.
Thanks!

(*Please note that this is a [fully automated](https://github.com/cvrebert/lmvtfy) comment.*)

**Comment by patrickhlauke** [comment]
@cvrebert @mdo good to merge, seeing that making an actual unit test appears to be elusive?

**Comment by patrickhlauke** [comment]
casting the net further... @hnrch02 @XhmikosR any thoughts?

**Comment by patrickhlauke** [comment]
As I haven't heard anything to the contrary, merged. If the absence of a unit test (which seems elusive, as the behavior seems to depend on actual browser interaction, which can't be tested in a straightforward way in a headless engine-only setup it seems) is a problem, I'd say let's revisit that aspect separately.

**Comment by alexw23** [comment]
:+1: for this issue I've managed to replicate the issue here http://codepen.io/anon/pen/MwwyRW

**Comment by alexw23** [comment]
I don't think this actually fixes is as now we have the issue that the focus still displays which is now even more confusing for the user.

**Comment by alexw23** [comment]
From what I can see the [one line](https://github.com/patrickhlauke/bootstrap/commit/1b9baa0d98019eba187e65102959fff60d06538a#diff-cb20e823e3cbd12f437d570630c85900L110) the other changes break the active class.

**Comment by cvrebert** [comment]
Your codepen uses v3.3.4's JS; the fix is in `master` (which will be v3.3.5)

**Comment by alexw23** [comment]
@cvrebert Is there a link to a master?

**Comment by cvrebert** [comment]
@alexw23 https://github.com/twbs/bootstrap/blob/master/dist/js/bootstrap.js

**Comment by alexw23** [comment]
@cvrebert As I mentioned these changes have broken it even further http://codepen.io/anon/pen/waaWGZ

**Comment by patrickhlauke** [comment]
@alexw23 your pen isn't working because of rawgithub

"Refused to execute script from 'https://raw.githubusercontent.com/twbs/bootstrap/master/dist/js/bootstrap.js' because its MIME type ('text/plain') is not executable, and strict MIME type checking is enabled.
index.html:3 Resource interpreted as Script but transferred with MIME type text/plain: "https://raw.githubusercontent.com/twbs/bootstrap/master/dist/js/bootstrap.js"."

**Comment by patrickhlauke** [comment]
However, I can see what you mean now about the class not being set once moving focus off the radio buttons (just brutally forked you pen and slammed bootstrap.js into the JS part) http://codepen.io/patrickhlauke/pen/zGGKwJ

**Comment by hnrch02** [comment]
(FYI: rawgit.com serves the raw GitHub files with the proper headers.)

**Comment by alexw23** [comment]
That seemed to be the issue I've updated the codepen. http://codepen.io/anon/pen/waaWGZ

However this doesn't work for enter key (for the first tab)

**Comment by patrickhlauke** [comment]
Note that, by default (talking native standard browser behavior here, without any Bootstrap) radio buttons don't really react to ENTER. Instead, if focusing on the first of a group of unselected radio buttons, you'd use SPACE to toggle it.

**Comment by alexw23** [comment]
Okay well I guess it's fixed then sorry for the confusion.

**Comment by patrickhlauke** [comment]
No need to apologize, and nope not quite fixed yet...seems the `.active` class doesn't stick correctly after selecting a radio button with the keyboard. The radio itself is correctly set, but `.active` doesn't seem to get actually set with my original fix (which i didn't notice because `.focus` c;ass still looks the same as `.active`). I'll work on a fix for that separately though.

**Comment by alexw23** [comment]
My diff fixes this however http://codepen.io/anon/pen/vOOKNj I guess you could remove the `keydown` and commit this.

**Comment by patrickhlauke** [comment]
But your diff seems to suffer from the same problem I bumped againsts: `change` event firing twice when using the keyboard, which is undesirable.

I *think* I managed to simplify the existing code and avoid any major pitfalls... http://codepen.io/patrickhlauke/pen/zGGEEY

**Comment by yuan3y** [comment]
@patrickhlauke seems like it fixes, finally! tried on chrome & firefox on linux.

**Comment by alexw23** [comment]
Just to be a pain we should look at replicating checkbox functionality also... I've added non-bootstrap radio/checkboxes for reference as to the normal usability. http://codepen.io/anon/pen/eNNGjO

**Comment by patrickhlauke** [comment]
More hacking around, seems to work now for checkboxes too (but the code is slowly getting a bit fugly, will need some cleaning up/reworking I think) http://codepen.io/patrickhlauke/pen/OVVOGL

**Pull request merged**